---
title: "Untitled"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Setup 

```{r}

"%&%" <- function(a,b) paste0(a,b)
library("data.table")
library("tidyverse")
library("GenomicRanges")

local.dir <- "/home/jason/science/projects/wtsa/joint_analyses/"
rescomp.dir <- "/home/jason/science/servers/FUSE5/"
cbrg.dir <- "/home/jason/science/servers/FUSE2/"
work.dir <- rescomp.dir %&% "projects/wtsa/joint_analyses/"
file.dir <- work.dir %&% "analysis_files/"
loc.df <- fread(file.dir %&% "locus_captures.txt")

```


```{r}

peakc.df <- fread(work.dir %&% "peakC_interactions/all_peakC-interactions_merge1K.txt")
deseq2.df <- fread(work.dir %&% "DESeq2_interactions/all_deseq2-interactions_merge1K.txt")

```


# Functions

```{r}

build_loc_inter_df <- function(){
  out.df <- c()
  no.inters <- c()
  pb <- txtProgressBar(min=0,max=dim(loc.df)[1],style=3)
  for (i in 1:dim(loc.df)[1]){
    setTxtProgressBar(pb,i)
    #print(i)
    row.df <- loc.df[i,]
    loc <- row.df$locus
    cap.vec <- (row.df$captures %>% strsplit(.,split=","))[[1]]
    for (e in 1:length(cap.vec)){
      cap <- cap.vec[e]
      exp <- ifelse(!grepl(x=cap,pattern="_"),"promoter",
                    ifelse(grepl(x=cap,pattern="__"),"enhancer_1st","enhancer_2nd"))
      cap.type <- ifelse(exp=="promoter","promoter",
                         ifelse(exp=="enhancer_2nd","credible.snp",
                                ifelse(grepl(x=cap,pattern="__uc__"),"upstream.control",
                                       ifelse(grepl(x=cap,pattern="__dc__"),"downstream.control","credible.snp"))))
      keep.names <- names(peakc.df)[names(peakc.df) %in% names(deseq2.df)]
      sub1 <- filter(peakc.df,capture==cap) %>% dplyr::select(.,one_of(c(keep.names,"padj")))
      sub2 <- filter(deseq2.df,capture==cap) %>% dplyr::select(.,one_of(keep.names))
      if (dim(sub1)[1]>0 & dim(sub2)[1]>0){
        names(sub1)[dim(sub1)[2]] <- "peakc.deseq2.padj"
        sub1$method <- "peakC"; sub2$method <- "DESeq2"
        sub2$peakc.deseq2.padj <- NA
        build.df <- rbind(sub1,sub2)
        build.df$locus <- loc
        out.df <- rbind(out.df,build.df)
      } else if(dim(sub1)[1]>0 & dim(sub2)[1]==0){
        names(sub1)[dim(sub1)[2]] <- "peakc.deseq2.padj"
        sub1$method <- "peakC"
        build.df <- sub1
        build.df$locus <- loc
        out.df <- rbind(out.df,build.df)
      } else if(dim(sub1)[1]==0 & dim(sub2)[1]>0){
        sub2$peakc.deseq2.padj <- NA
        sub2$method <- "DESeq2"
        build.df <- sub2   
        build.df$locus <- loc
        out.df <- rbind(out.df,build.df)
      } else{
        print("No interactions available for capture: " %&% cap %&% ",at locus: " %&% loc)
        no.inters <- append(no.inters,cap)
      }
    }    
  }
  return(list(out.df,no.inters))
}

```


# Execution 

```{r}

loc.list <- build_loc_inter_df()
full.df <- loc.list[[1]]
no.inters <- loc.list[[2]]

write.table(x=full.df,file=file.dir%&%"all_locus_interactions_merge1K.txt",sep="\t",row.names=F,quote=F)
write.table(x=no.inters,file=file.dir%&%"captures_without_interactions_merge1K.txt",sep="\t",row.names=F,quote=F,
            col.names=F)

```


```{r}

loc.df$locus[!(loc.df$locus %in% unique(full.df$locus))]
full.df$method %>% table(.) # DESeq2 = 1177, peakC = 1156
full.df$type %>% table(.) # distal = 788, proximal = 1545
(filter(full.df,method=="peakC",peakc.deseq2.padj<0.05) %>% dim(.))[1] # 546
all.genes <- map(full.df$genes,function(g){
  strsplit(g,split=",")
}) %>% unlist(.) %>% unique(.)

```

Note now 2,333 total interactions, was previously 2,315 when used DEseq2 (method 2) range of 500kb rather than 1Mb 

```{r}

peakc.gr <- GRanges(seqnames=filter(full.df,method=="peakC")$seqnames,
                    IRanges(filter(full.df,method=="peakC")$start,filter(full.df,method=="peakC")$end))
deseq2.gr <- GRanges(seqnames=filter(full.df,method=="DESeq2")$seqnames,
                    IRanges(filter(full.df,method=="DESeq2")$start,filter(full.df,method=="DESeq2")$end))
all.gr <- append(peakc.gr,deseq2.gr)

# 2333 total interactions, up from previous total of 2315
# 1156 peakC interactions
# 1177 DESeq2 interactions, increased from 1159 DESeq2 interactions with previous range (500kb)

all.red.gr <- reduce(all.gr) # 1310 non-overlapping regions (previously 1299)
peakc.red.gr <- reduce(peakc.gr) # 735 non-overlapping regions
deseq2.red.gr <- reduce(deseq2.gr) #  830 non-overlapping regions (previously 823)

ov1 <- findOverlapPairs(peakc.gr,deseq2.gr) %>% as.data.frame(.)
ov1$name <- ov1$first.seqnames %&% ":" %&% ov1$first.start %&% "-" %&% ov1$first.end
unique(ov1$name) %>% unique(.) %>% length(.) #  354 (30.6%) peakC interactions overlap DESeq2 interactions; previously 333

ov2 <- findOverlapPairs(deseq2.gr,peakc.gr) %>% as.data.frame(.)
ov2$name <- ov2$first.seqnames %&% ":" %&% ov2$first.start %&% "-" %&% ov2$first.end
unique(ov2$name) %>% unique(.) %>% length(.) # 360 (30.6%) DESeq2 interactions overlap peakC interactions; previously 364

ov1b <- findOverlapPairs(peakc.red.gr,deseq2.red.gr) %>% as.data.frame(.)
ov1b$name <- ov1b$first.seqnames %&% ":" %&% ov1b$first.start %&% "-" %&% ov1b$first.end
unique(ov1b$name) %>% unique(.) %>% length(.) # 183 (24.9%) non-overlapping peakC interactions interact non-overlapping DESeq2 interactions ; previously 179

ov2b <- findOverlapPairs(deseq2.red.gr,peakc.red.gr) %>% as.data.frame(.)
ov2b$name <- ov2b$first.seqnames %&% ":" %&% ov2b$first.start %&% "-" %&% ov2b$first.end
unique(ov2b$name) %>% unique(.) %>% length(.) # 240 (28.9%) non-overlapping DESeq2 interactions interact non-overlapping peakC interactions ; previously 244

```


```{r}

full.df$type %>% table(.) # distal = 788, proximal = 1545
filter(full.df,method=="peakC")$type %>% table(.) # distal = 458, proximal = 698
filter(full.df,method=="DESeq2")$type %>% table(.) # distal = 330, proximal = 847
filter(full.df,method=="peakC",peakc.deseq2.padj<0.05)$type %>% table(.) # distal = 197, proximal = 349 
```


Summary notes: 

Took an inventory of the interactions detected within each of the 130 locus regions. From this, found a total of 2,333 interactions across 128 locus regions 

The two loci without any capture interactions are locus_35 (MRPS30, Enh2ndRound) and locus_53 (IGF2BP3, Enh2ndRound)

Of the 2,333 interactions (1,310 non-overlapping regions), 1,177 (50.4%) were differentially present in Endos (i.e. DESeq2) and 1,156 (49.5%) were significant in peakC

Of the 1,156 peakC interactions, 354 (30.6%) overlap with DESeq2 interactions 
Of the 1,177 DESeq2 interactions, 360 (30.6%) overlap with peakC interactions: Can confirm that the double 30.6% is not a mistake 

Across captures, the peakC interactions merge into 735 non-overlapping regions, of which, 183 (24.9%) overlap non-overlapping DESeq2 interactions 
Across captures, the DESeq2 interactions merge into 830 non-overlapping regions, of which, 240 (28.9%) overlap non-overlapping DESeq2 interactions 

Of the 1,156 peakC interactions, 546 (47.2% of peakC interactions) were differentially open from DESeq2 (padj<0.05): Note that this analysis was peformed across all peakC interactions first and different from DESeq2 performed at all individual fragments within 1Mb window) 

of the 2,333 total interactions, 1,545 (66.2%) were proximal (i.e. within 250kb of capture site) and 788 (33.8%) were distal (beyond 250kb) 







