---
title: "08.2.1"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Setup 

```{r}
"%&%" <- function(a,b) paste0(a,b)
library("data.table")
library("tidyverse")
library("GenomicRanges")

rescomp.dir <- "/Users/jasont/science/servers/FUSE5/"
work.dir <- rescomp.dir %&% "projects/wtsa/joint_analyses/"
output.dir <- work.dir %&%  "analysis_files/master_tables/"
credsets.df <- fread(output.dir %&% "t2d-credible-sets-table.txt")
credsnps.df <- fread(output.dir %&% "t2d-credible-snps-table.txt")
```


## Append SNP info to eQTL colocalisation data frame  

```{r}
qtl.dir <- work.dir %&% "coloc_analysis_files/summary_files/"
coloc.snp.df <- fread(qtl.dir %&% "coloc_t2d-gwas-islet-eqtl_sig-p50.txt")
coloc.sum.df <- fread(qtl.dir %&% "coloc_t2d-gwas-islet-eqtl_summary.txt")

```


```{r}

names(coloc.snp.df)[4] <- "SNPID"
pb <- txtProgressBar(min=0,max=dim(coloc.snp.df)[1],style=3)
coloc.snp.df$Signal <- purrr::map(1:dim(coloc.snp.df)[1],function(i){
  setTxtProgressBar(pb,i)
  s <- coloc.snp.df$gwas.signal[i]
  strsplit(x=s,split="-")[[1]][2]
}) %>% as.character(.)

pb <- txtProgressBar(min=0,max=dim(coloc.sum.df)[1],style=3)
coloc.sum.df$Signal <- purrr::map(1:dim(coloc.sum.df)[1],function(i){
  setTxtProgressBar(pb,i)
  s <- coloc.sum.df$gwas.signal[i]
  strsplit(x=s,split="-")[[1]][2]
}) %>% as.character(.)
```


## Append eQTL colocalisation info to credible set/SNP tables 

```{r}

pb <- txtProgressBar(min=0,max=dim(credsets.df)[1],style=3)
for (i in 1:dim(credsets.df)[1]){
  setTxtProgressBar(pb,i)
  row.df <- credsets.df[i,]
}

```

