---
title: "Untitled"
author: "Jason Torres"
date: "5/9/2019"
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Setup 

```{r}

"%&%" <- function(a,b) paste0(a,b)
library("data.table")
library("tidyverse")
library("ggbio")
library("RColorBrewer")
library("viridis")
library(Homo.sapiens)

rescomp.dir <- "/home/jason/science/servers/FUSE5/" # "/Users/jtorres/FUSE5/" 
cbrg.dir <-  "/home/jason/science/servers/FUSE2/" # "/Users/jtorres/FUSE2/"
got2d.dir <-  "/home/jason/science/servers/FUSE/" # "/Users/jtorres/FUSE/"

enh1.dir <- rescomp.dir %&% "projects/wtsa/enhancer_driven/first_round/statistical_anlayses/output_files/"
prom.dir <- rescomp.dir %&% "projects/wtsa/promoter_driven/statistical_analysis/output_files/"

enh1.mod.df <- fread(enh1.dir %&% "peakC-modeled-default_EndoC_withBaits.txt")
enh1.mod.df$inter.type <- "modeled"
prom.mod.df <- fread(prom.dir %&% "peakC-modeled-default_EndoC_withBaits.txt")
prom.mod.df$inter.type <- "modeled"

enh1.comp.df <- fread(enh1.dir %&% "peakC-comparative_withBaits.txt")
enh1.comp.df$inter.type <- "compared"
prom.comp.df <- fread(prom.dir %&% "peakC-comparative_withBaits.txt")
prom.comp.df$inter.type <- "compared"

prom.df <- rbind(prom.mod.df,prom.comp.df)
enh1.df <- rbind(enh1.mod.df,enh1.comp.df)


peak.to.genes <- fread(enh1.dir %&% "peaks-to-genes_cv_eqtlCheck_all.txt")

```


# Profile interaction features 

```{r}

library("GenomicRanges")

build_interaction_df <- function(df){
  names(df)[1] <- "capture"
  out.df <- c()
  sub <- filter(df,type=="bait")[,1:5]
  sub <- sub[!duplicated(sub),]
  for (i in 1:dim(sub)[1]){
    row.df <- sub[i,1:5]
    names(row.df) <- c("capture","chrom","bait.start","bait.end","bait.width")
    cap <- row.df$capture
    sub2 <- filter(df,capture==cap,type=="interaction")
    for (i in 1:dim(sub2)[1]){
      row.df2 <- sub2[i,c(3:5,7)]
      names(row.df2) <- c("inter.start","inter.end","inter.width","inter.type")
      r.df <- cbind(row.df,row.df2)
      out.df <- rbind(out.df,r.df)
    }
  }
  inter.dist <- map(1:dim(out.df)[1],function(i){
    row.df <- out.df[i,]
    bait.gr <- GRanges(seqnames = row.df$chrom,IRanges(row.df$bait.start,row.df$bait.end))
    inter.gr <- GRanges(seqnames = row.df$chrom,IRanges(row.df$inter.start,row.df$inter.end))
    distance(bait.gr,inter.gr)
  }) %>% as.integer(.)
  out.df$inter.dist <- inter.dist
  return(out.df)
}


build_summary_df <- function(df){
  int.df <- build_interaction_df(df)
  cap.vec <- int.df$capture %>% unique(.)
  out.df <- c()
  for (cap in cap.vec){
    sub <- filter(int.df,capture==cap)
    build.df <- data.frame("capture"=cap,"chrom"=sub$chrom[1],"num.inter"=dim(sub)[1],
                           "num.mod"=sum(sub$inter.type=="modeled"),
                           "num.comp"=sum(sub$inter.type=="compared"),
                           "med.inter.width"=median(sub$inter.width), "min.inter.dist"=min(sub$inter.dist),
                           "med.inter.dist"=median(sub$inter.dist),"max.inter.dist"=max(sub$inter.dist),stringsAsFactors = F)
    out.df <- rbind(out.df,build.df)
  }
  return(out.df)
}


```



# Promoter capture 

```{r}

sum.df1 <- build_summary_df(prom.df) 
summary(sum.df1$num.inter)
write.table(x=sum.df1,file=prom.dir %&% "peakC-interaction-summary.txt",
            sep="\t",quote=F,row.names=F)

```

There are `r dim(sum.df1)[1]` captures with at least one significant peakC interaction 

The total number of peakC interactions from this experiment is  `r sum.df1$num.inter %>% sum(.)`
The total number of "modeled" peakC interactions from this experiment is  `r sum.df1$num.mod %>% sum(.)`
The total number of "comparative" peakC interactions from this experiment is  `r sum.df1$num.comp %>% sum(.)`
The proportion of "modeled" peakC interactions from this experiment is  `r (sum.df1$num.mod %>% sum(.))/(sum.df1$num.inter %>% sum(.))`


The median number of interactions at a capture site is `r median(sum.df1$num.inter)` 
The minimum number of interactions at a capture site is `r min(sum.df1$num.inter)` and correponds to `r arrange(sum.df1,num.inter)[1,]$capture` 
The maximum number of interactions at a capture site is `r max(sum.df1$num.inter)` and correponds to `r arrange(sum.df1,-(num.inter))[1,]$capture` 

The median interaction width at a capture site is `r median(sum.df1$med.inter.width)` 
The minimum interaction width at a capture site is `r min(sum.df1$med.inter.width)` and correponds to `r arrange(sum.df1,med.inter.width)[1,]$capture`
The maximum interaction width at a capture site is `r max(sum.df1$med.inter.width)` and correponds to `r arrange(sum.df1,-med.inter.width)[1,]$capture`

The median interaction distance at a capture site is `r median(sum.df1$med.inter.dist)` 
The minimum interaction distance at a capture site is `r min(sum.df1$min.inter.dist)` and correponds to `r arrange(sum.df1,min.inter.dist)[1,]$capture`
The maximum interaction distance at a capture site is `r max(sum.df1$max.inter.dist)` and correponds to `r arrange(sum.df1,-max.inter.dist)[1,]$capture`




# Enhancer capture (1st)

```{r}

sum.df2 <- build_summary_df(enh1.df) 
summary(sum.df2$num.inter)
write.table(x=sum.df2,file=enh1.dir %&% "peakC-interaction-summary.txt",
            sep="\t",quote=F,row.names=F)

```

There are `r dim(sum.df2)[1]` captures with at least one significant peakC interaction 

The total number of peakC interactions from this experiment is  `r sum.df2$num.inter %>% sum(.)`
The total number of "modeled" peakC interactions from this experiment is  `r sum.df2$num.mod %>% sum(.)`
The total number of "comparative" peakC interactions from this experiment is  `r sum.df2$num.comp %>% sum(.)`
The proportion of "modeled" peakC interactions from this experiment is  `r (sum.df2$num.mod %>% sum(.))/(sum.df2$num.inter %>% sum(.))`


The median number of interactions at a capture site is `r median(sum.df2$num.inter)` 
The minimum number of interactions at a capture site is `r min(sum.df2$num.inter)` and correponds to `r arrange(sum.df2,num.inter)[1,]$capture` 
The maximum number of interactions at a capture site is `r max(sum.df2$num.inter)` and correponds to `r arrange(sum.df2,-(num.inter))[1,]$capture` 

The median interaction width at a capture site is `r median(sum.df2$med.inter.width)` 
The minimum interaction width at a capture site is `r min(sum.df2$med.inter.width)` and correponds to `r arrange(sum.df2,med.inter.width)[1,]$capture`
The maximum interaction width at a capture site is `r max(sum.df2$med.inter.width)` and correponds to `r arrange(sum.df2,-med.inter.width)[1,]$capture`

The median interaction distance at a capture site is `r median(sum.df2$med.inter.dist)` 
The minimum interaction distance at a capture site is `r min(sum.df2$min.inter.dist)` and correponds to `r arrange(sum.df2,min.inter.dist)[1,]$capture`
The maximum interaction distance at a capture site is `r max(sum.df2$max.inter.dist)` and correponds to `r arrange(sum.df2,-max.inter.dist)[1,]$capture`



```{r}

int.df1 <- build_interaction_df(prom.df)
int.df2 <- build_interaction_df(enh1.df)

write.table(x=int.df1,file=prom.dir %&% "peakC-interaction_loops.txt",
            sep="\t",quote=F,row.names=F)
write.table(x=int.df2,file=enh1.dir %&% "peakC-interaction_loops.txt",
            sep="\t",quote=F,row.names=F)


```


# Append peak to gene info for interaction loops 


```{r}

append_gene_to_loopdf <- function(loop.df){
  out.df <- c()
  for (i in 1:dim(loop.df)[1]){
    row.df <- loop.df[i,]
    sub <- filter(peak.to.genes,variant==row.df$capture,start==row.df$inter.start,end==row.df$inter.end) %>% 
      dplyr::select(.,one_of("ensGene","eqtl.support"))
    sub <- sub[!duplicated(sub),]
    num.genes <- dim(sub)[1]
    genes <- sub$ensGene %>% paste0(.,collapse=",")
    eqtl.support <- any(sub$eqtl.support==TRUE)  
    row.df$num.genes <- num.genes; row.df$genes <- genes
    row.df$eqtl.support <- eqtl.support
    out.df <- rbind(out.df,row.df)
  }  
  return(out.df)
}


append_gene_to_summary <- function(loop.df,sum.df){
  out.df <- c()
  for (i in 1:dim(sum.df)[1]){
    row.df <- sum.df[i,]
    sub <- filter(loop.df,capture==row.df$capture)
    num.gene.inter <- sub$num.genes %>% sum(.)
    vec <- (map(sub$genes,function(s){nchar(s)}) %>% as.integer(.)) > 0
    genes <- sub$genes[vec] %>% paste0(.,collapse=",")
    eqtl.support <- sub$eqtl.support %>% sum(.)
    row.df$num.gene.inter <- num.gene.inter
    row.df$genes <- genes
    row.df$eqtl.support <- eqtl.support
    out.df <- rbind(out.df,row.df)
  }
  return(out.df)
}


```

```{r}

loop.df <- int.df2
sum.df <- sum.df2
loop.df <- append_gene_to_loopdf(loop.df)
sum.df <- append_gene_to_summary(loop.df,sum.df)

status <- map(sum.df$capture,function(cap){
  if((grepl(x=cap,pattern="__uc__") | grepl(x=cap,pattern="__dc__"))){
    stat <- "Control.SNP"
  } else{
    stat <-"T2D.SNP"
  }
  return(stat)
}) %>% as.character(.)

sum.df$status <- status 


write.table(x=loop.df,file=prom.dir %&% "peakC-interaction_loops_withGeneInfo.txt",
            sep="\t",quote=F,row.names=F)
write.table(x=sum.df,file=enh1.dir %&% "peakC-interaction-summary_withGeneInfo.txt",
            sep="\t",quote=F,row.names=F)
```


There are `r sum.df$num.gene.inter %>% median(.)` interactions with gene TSS per capture 
There are `r sum.df$num.gene.inter %>% median(.)` interactions with gene TSS per capture 



Append SNP PPA 

```{r}


cred.df <- fread(rescomp.dir %&% "projects/t2d_classification/revamp/genetic_credible_sets/gencred.txt")

s <- "PTGFRN__4DF__cv__chr1_117529458"

sum.df$SNP.PPA <- map(sum.df$capture,function(s){
  vec <- (strsplit(s,split="__")[[1]][4] %>% strsplit(.,split="_"))[[1]]
  filter(cred.df,CHR==vec[1],POS==vec[2])$PPA %>% max(.)
}) %>% as.numeric(.)





```



# Compare controls vs T2D SNPs 


```{r}


boxplot(data=sum.df,num.inter ~ status)

boxplot(filter(sum.df,status=="Control.SNP")$num.inter,
        filter(sum.df,SNP.PPA>0.50,status=="T2D.SNP")$num.inter)

#boxplot(data=sum.df,eqtl.support ~ status)

hist(filter(sum.df,status=="Control.SNP")$num.inter)
median(filter(sum.df,status=="Control.SNP")$num.inter)

hist(filter(sum.df,status=="T2D.SNP")$num.inter)
median(filter(sum.df,status=="T2D.SNP")$num.inter)

wilcox.test(filter(sum.df,status=="Control.SNP")$num.inter,
       filter(sum.df,status=="T2D.SNP")$num.inter) # P-value 0.15
wilcox.test(filter(sum.df,status=="Control.SNP")$num.inter,
        filter(sum.df,SNP.PPA>0.50,status=="T2D.SNP")$num.inter) # P-value = 0.7899
wilcox.test(filter(sum.df,status=="Control.SNP")$eqtl.support,
        filter(sum.df,SNP.PPA>0.50,status=="T2D.SNP")$eqtl.support) # P-value = 0.6151

```


T2D Captures show no difference in number of significant interactions 


```{r}

hist(filter(sum.df,status=="Control.SNP")$num.gene.inter)
median(filter(sum.df,status=="Control.SNP")$num.gene.inter)

hist(filter(sum.df,status=="T2D.SNP")$num.gene.inter)
median(filter(sum.df,status=="T2D.SNP")$num.gene.inter)

wilcox.test(filter(sum.df,status=="Control.SNP")$num.gene.inter,
       filter(sum.df,status=="T2D.SNP")$num.gene.inter) # P-value 0.1246

```

T2D Captures show no difference in number of significant interactions with gene TSS 



```{r}

hist(filter(sum.df,status=="Control.SNP")$eqtl.support)
hist(filter(sum.df,status=="T2D.SNP")$eqtl.support)
wilcox.test(filter(sum.df,status=="Control.SNP")$eqtl.support,
       filter(sum.df,status=="T2D.SNP")$eqtl.support) # 

```

T2D Captures show no difference in number of significant interactions with gene TSS 


```{r}

gencred.df <- fread(rescomp.dir %&% "datasets/diamante_hrc/gencred.txt")

```


```{r}

lookupCap <- function(g){
  ensg <- filter(peak.to.genes,Gene.name==g)$ensGene %>% unique(.)
  filter(loop.df,grepl(ensg,genes))$capture  
}

lookupCap("FBRSL1")
#"FBRSL1__109DF__cv__chr12_133066392" # PPA = 0.13 *** 

lookupCap("SLC38A11")
#[1] "COBLL1__16DF__cv__chr2_165545615" PPA = 0.009
#"COBLL1__16DF__cv__chr2_165548569" PPA = 0.012
#[3] "COBLL1__16DF__cv__chr2_165554302" PPA = 0.017
#"COBLL1__16DF__dc__chr2_165628430"

lookupCap("DDX51")
#[1] "FBRSL1__109DF__cv__chr12_133068484" PPA = 0.16
#"FBRSL1__109DF__cv__chr12_133069698" PPA = 0.18 (index) ***
#[3] "FBRSL1__109DF__cv__chr12_133072767" PPA = NA
#"FBRSL1__109DF__cv__chr12_133080449" PPA = 0.16
#[5] "FBRSL1__109DF__cv__chr12_133087707" PPA = 0.02

lookupCap("PDE8B")


```

