---
title: "02.1_Prepare-islet-eQTL-df.Rmd"
author: "Jason Torres"
date: "February 15, 2017"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


Note: This data was generated by fastQTL. It contains gene-level eQTL results from 426 INSPIRE islet samples. 


```{r}
"%&%" <- function(a,b) paste0(a,b) 
library("data.table")
library("dplyr")
#source("https://bioconductor.org/biocLite.R")
#biocLite("EnsDb.Hsapiens.v75")
library("EnsDb.Hsapiens.v75")

serv.dir <- "/Users/jtorres/FUSE/"
rds.dir <- serv.dir %&% "projects/wtsa/enhancer_driven/probe_design/rds/"

```


Islet eQTLs, FDR 1% of nominal p-values 

```{r}

eqtl.file <- serv.dir%&%"reference/islet/eqtls/inspire/nominal_pass/output/" %&% "eqtls_fdr05.txt.gz"
key.file <- serv.dir%&%"reference/islet/eqtls/inspire/snp_keyfile.txt.gz"

e.df <- fread("cat " %&% eqtl.file %&% " | zmore")
names(e.df) <- c("ENSID","RSID","DIST","P","Beta","Q") #c("ENSID","SNPs.Tested","MLE1","MLE2","DUMMY","RSID","DIST","NOM.P","Beta","P1","P")

#e.df$Q <- p.adjust(e.df$P, method="fdr")


key.df <- fread("cat " %&% key.file %&% " | zmore") 
key.df$POS <- gsub("31948303_A/G","31948303",key.df$POS) # manual adjustement based on output from fread warning 
key.df$POS <- as.integer(key.df$POS)
#names(e.df) <- c("RSID","ENSID","Beta","T","P","Q")
e.df <- inner_join(e.df,key.df,by="RSID")
e.df <- e.df[!duplicated(e.df),]

ensref.df <- data.frame(genes(EnsDb.Hsapiens.v75)$gene_id,
                        genes(EnsDb.Hsapiens.v75)$gene_name,
                        stringsAsFactors = FALSE)
names(ensref.df) <- c("ENSID","GENE")

ensid2gene <- function(){
  pb <- txtProgressBar(min = 0, max = dim(e.df)[1], initial = 0, style = 3)
  GENE <- as.character(sapply(1:length(e.df$ENSID), function(i){
    setTxtProgressBar(pb, i) 
    e <- e.df$ENSID[i]
    name <- strsplit(x=e,split=".",fixed=TRUE)[[1]][1]
    gene <- filter(ensref.df,ENSID==name)$GENE
    return(gene)
  }))
  close(pb)
  return(GENE)
}

GENE <- ensid2gene()
e.df <- cbind(e.df,GENE); e.df$GENE <- as.character(e.df$GENE)

SNPID <- as.character(sapply(1:length(e.df$CHR),function(i){
  s <- paste(e.df$CHR[i],e.df$POS[i],sep=":")
  return(s)
}))
e.df <- cbind(e.df,SNPID)  
e.df$SNPID <- as.character(e.df$SNPID)
eqtl.df <- e.df 
fdr1.df <- filter(eqtl.df,Q<=0.01)
saveRDS(fdr1.df,file=rds.dir%&%"eqtl.fdr01.df.RDS")

```


Islet eQTLs, Top eQTLs per gene (permutations) 

```{r}

eqtl.file <- serv.dir%&%"reference/islet/eqtls/inspire/output/" %&% "permutations.all.chunks.txt.gz"
key.file <- serv.dir%&%"reference/islet/eqtls/inspire/output/snp_keyfile.txt.gz"

e.df <- fread("cat " %&% eqtl.file %&% " | zmore")
names(e.df) <- c("ENSID","SNPs.Tested","MLE1","MLE2","DUMMY","RSID","DIST","NOM.P","Beta","P1","P")

e.df$Q <- p.adjust(e.df$P, method="fdr")


key.df <- fread("cat " %&% key.file %&% " | zmore") 
key.df$POS <- gsub("29324054_A/G","29324054",key.df$POS) # manual adjustement based on output from fread warning 
key.df$POS <- as.integer(key.df$POS)
#names(e.df) <- c("RSID","ENSID","Beta","T","P","Q")
e.df <- inner_join(e.df,key.df,by="RSID")
e.df <- e.df[!duplicated(e.df),]

ensref.df <- data.frame(genes(EnsDb.Hsapiens.v75)$gene_id,
                        genes(EnsDb.Hsapiens.v75)$gene_name,
                        stringsAsFactors = FALSE)
names(ensref.df) <- c("ENSID","GENE")

ensid2gene <- function(){
  pb <- txtProgressBar(min = 0, max = dim(e.df)[1], initial = 0, style = 3)
  GENE <- as.character(sapply(1:length(e.df$ENSID), function(i){
    setTxtProgressBar(pb, i) 
    e <- e.df$ENSID[i]
    name <- strsplit(x=e,split=".",fixed=TRUE)[[1]][1]
    gene <- filter(ensref.df,ENSID==name)$GENE
    return(gene)
  }))
  close(pb)
  return(GENE)
}

GENE <- ensid2gene()
e.df <- cbind(e.df,GENE); e.df$GENE <- as.character(e.df$GENE)

SNPID <- as.character(sapply(1:length(e.df$CHR),function(i){
  s <- paste(e.df$CHR[i],e.df$POS[i],sep=":")
  return(s)
}))
e.df <- cbind(e.df,SNPID)  
e.df$SNPID <- as.character(e.df$SNPID)
eqtl.df <- e.df 
saveRDS(eqtl.df,file=rds.dir%&%"eqtl.perm.df.RDS")

```


