---
title: "Sensitivity analysis"
author: "Jason Torres"
date: "August 18, 2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# Setup 

```{r}

"%&%" <- function(a,b) paste0(a,b)
library("dplyr")
library("data.table")
library("ggplot2")
library("gridExtra")
library("GenomicRanges")

serv.dir <- "/Users/jtorres/FUSE2/" # FUSE2
stat.dir <- serv.dir %&% "analysis/wtsa/promoter_driven/statistical_analysis/"
ref.dir <- serv.dir %&% "analysis/wtsa/promoter_driven/reference_files/"
out.dir <- stat.dir %&% "output_files/"
res.dir <- serv.dir %&% "promoter-driven/Stat_Package_CB4/Processed_gfc/R_files/"
eda.dir <- serv.dir %&% "analysis/wtsa/promoter_driven/EDA/"

```


Extract Functions 

```{r}



get_subtract <- function(genename,e.gr){
  #e.df <- filter(eqtl.df,grepl(genename,eqtl.df$V4))
  #e.gr <- GRanges(seqnames=e.df$V1,IRanges(e.df$V2,e.df$V3))
  temp.df <- suppressWarnings(fread(res.dir %&% genename %&% "_Processed_R_analysis.txt"))
  names(temp.df) <- c("id","Processed_gffsForCisOnly.EndoB_A",	
                      "Processed_gffsForCisOnly.EndoB_C",	"Processed_gffsForCisOnly.EndoB_D",
                      "Processed_gffsForCisOnly.Blymph_A",
                      "Processed_gffsForCisOnly.Blymph_B",
                      "Processed_gffsForCisOnly.Blymph_C",
                      "conD_1_norm",	"conD_2_norm",	"conD_3_norm",	"conT_1_norm",
                      "conT_2_norm",	"conT_3_norm",	"conD_mean",	"conT_mean",
                      "DIF_conT_minus_conD_mean",	"DIF_conD_minus_conT_mean",	"conD_SD",
                      "conT_SD",	"conD_CV",	"conT_CV",	"conD_signoise",	"conT_signoise")
  chromo <- sapply(temp.df$id, function(str){
    strsplit(x=str,split=":")[[1]][1]})
  start <- sapply(temp.df$id, function(str){ 
    as.integer(strsplit(strsplit(x=str,split="-")[[1]][1],split=":")[[1]][2])})
  end <- sapply(temp.df$id,function(str){
    as.integer(strsplit(x=str,split="-")[[1]][2])})
  gr <- GRanges(seqnames=chromo,IRanges(start,end))
  out.gr <- gr[gr %over% e.gr]
  keep.chromo <- as.character(seqnames(out.gr))
  keep.start <- start(ranges(out.gr))
  keep.end <- end(ranges(out.gr))
  keep.id <- keep.chromo %&% ":" %&% keep.start %&% "-" %&% keep.end
  out.df <- filter(temp.df,id %in% keep.id) %>% # id == keep.id for single 
    select(one_of("id","conD_mean","conT_mean","DIF_conD_minus_conT_mean"))
  return(out.df)
}

get_deseq <- function(genename,e.gr){
  temp.df <- suppressWarnings(fread(res.dir %&% genename %&% "_Processed_DESeq_analysis.txt"))
  names(temp.df) <- c("id","baseMean","log2FoldChange","lfcSE","stat",
                      "pvalue","padj","logpvalue","logpadj")
  chromo <- sapply(temp.df$id, function(str){
    strsplit(x=str,split=":")[[1]][1]})
  start <- sapply(temp.df$id, function(str){ 
    as.integer(strsplit(strsplit(x=str,split="-")[[1]][1],split=":")[[1]][2])})
  end <- sapply(temp.df$id,function(str){
    as.integer(strsplit(x=str,split="-")[[1]][2])})
  gr <- GRanges(seqnames=chromo,IRanges(start,end))
  out.gr <- gr[gr %over% e.gr]
  keep.chromo <- as.character(seqnames(out.gr))
  keep.start <- start(ranges(out.gr))
  keep.end <- end(ranges(out.gr))
  keep.id <- keep.chromo %&% ":" %&% keep.start %&% "-" %&% keep.end
  out.df <- filter(temp.df,id %in% keep.id) %>% 
    select(one_of("id","log2FoldChange","pvalue","padj","logpadj"))
  names(out.df) <- c("id","deseq.l2fc","deseq.pval","deseq.padj","deseq.logpadj")
  return(out.df)
}

get_peakC_mod <- function(e.gr,win,fdr.thresh){
  fdr.char <- as.character(fdr.thresh)
  if (fdr.thresh==0.10){
    fdr.char <- "0.10"
  }
  if (fdr.thresh==0.20){
    fdr.char <- "0.20"
  }
  fname <- out.dir %&% "peakC-modeled_fdr"%&%fdr.char%&%"_win"%&%win%&%".txt"
  #print(fname)
  temp.df <- fread(fname)
  gr <- GRanges(seqnames=temp.df$chr,mcols=temp.df$capture,
                IRanges(start=temp.df$frag.start,end=temp.df$frag.end)) 
  out.gr <- gr[gr %over% e.gr]
  keep.chromo <- as.character(seqnames(out.gr))
  keep.start <- start(ranges(out.gr))
  keep.end <- end(ranges(out.gr))
  keep.id <- keep.chromo %&% ":" %&% keep.start %&% "-" %&% keep.end
  #keep.id <- ifelse(keep.id==":-",FALSE,keep.id)
  keep.id <- ifelse(keep.id==":-",FALSE,TRUE)
  return(keep.id)
}

get_peakC_comp <- function(e.gr,win,abscutoff){
  fname <- out.dir %&% "peakC-comparative_abscutoff-"%&%abscutoff%&%"_win"%&%win%&%".txt"
  temp.df <- fread(fname)
  gr <- GRanges(seqnames=temp.df$chr,mcols=temp.df$capture,
                IRanges(start=temp.df$frag.start,end=temp.df$frag.end)) 
  out.gr <- gr[gr %over% e.gr]
  keep.chromo <- as.character(seqnames(out.gr))
  keep.start <- start(ranges(out.gr))
  keep.end <- end(ranges(out.gr))
  keep.id <- keep.chromo %&% ":" %&% keep.start %&% "-" %&% keep.end
  #keep.id <- ifelse(keep.id==":-",FALSE,keep.id)
  keep.id <- ifelse(keep.id==":-",FALSE,TRUE)
  return(keep.id)
}

```


# Sensitivity analysis 

```{r}


build_sensitivity_summary <- function(){
  eqtl.df <- as.data.frame(fread(ref.dir %&% "eQTL-index.bed"))
  eqtl.df <- eqtl.df[!duplicated(eqtl.df),]
  eqtl.df$V2 <- (eqtl.df$V3 - 1 ) 
  eqtl.df$V2 <- as.integer(eqtl.df$V2)
  out.df <- c() 
  pb <- txtProgressBar(min=0,max=dim(eqtl.df)[1],style=3)
  for (i in 1:dim(eqtl.df)[1]){
    #print(i) 
    setTxtProgressBar(pb,i)
    sub.df <- eqtl.df[i,]
    e.gr <- GRanges(seqnames=sub.df$V1,
                    IRanges(sub.df$V2,sub.df$V3),
                    mcols=sub.df$V4)
    gene <- strsplit(x=sub.df$V4,split="_")[[1]][2]
    df1 <- get_subtract(gene,e.gr)
    df2 <- get_deseq(gene,e.gr)
    df2$deseq.sig <- (df2$deseq.padj > -log(0.05,base=exp(1)))
    df <- inner_join(df1,df2,by="id")
    df$peakCmod.win11.fdr05 <- get_peakC_mod(e.gr,11,0.05)
    df$peakCmod.win60.fdr05 <- get_peakC_mod(e.gr,60,0.05)
    df$peakCcomp.win11.cutoff140 <- any(get_peakC_comp(e.gr,11,140)==TRUE)
    df$peakCcomp.win60.cutoff140 <- any(get_peakC_comp(e.gr,60,140)==TRUE)
    df$peakCcomp.win11.cutoff10 <- any(get_peakC_comp(e.gr,11,10)==TRUE)
    df$peakCcomp.win60.cutoff10 <- any(get_peakC_comp(e.gr,60,10)==TRUE)
    names(sub.df) <- c("chrom","pos0","pos","eqtl")
    df <- cbind(select(sub.df,one_of("chrom","pos","eqtl")),df)
    names(df)[grepl("id",names(df))] <- "frag.id"
    out.df <- rbind(out.df,df)
  }
  return(out.df)
}

sens.df <- build_sensitivity_summary()

write.table(x=sens.df,file=stat.dir %&% "sensitivity-table-updated.txt",
            sep="\t",row.names=FALSE,quote=FALSE)


get_sensitivity_significant <- function(sens.df){
  sig.vec <- c()
  for (i in 1:dim(sens.df)[1]){
    print(i)
    sig.vec <- append(sig.vec,any(sens.df[i,13:18]==TRUE))
    #sig.vec <- append(sig.vec,any(sens.df[i,13:16]==TRUE))

  }
  sens <- dim(sens.df[sig.vec,])[1] / dim(sens.df)[1] # sensitivity 36%   
  print("Sensitivity: " %&% round(sens,3))
  sens.sig.df <- select(sens.df[sig.vec,],
                        one_of("chrom","pos","eqtl",
                                "frag.id","conD_mean","conT_mean",
                                "DIF_conD_minus_conT_mean",
                                "deseq.sig","peakCmod.win11.fdr05",
                                "peakCmod.win60.fdr05",
                                "peakCcomp.win11.cutoff140",
                                "peakCcomp.win60.cutoff140",
                                "peakCcomp.win11.cutoff10",
                                "peakCcomp.win60.cutoff10"))
  return(sens.sig.df)
}

sens.sig.df <- get_sensitivity_significant(sens.df)
write.table(x=sens.sig.df,file=stat.dir %&% "sensitivity-sig-table-updated.txt",
            sep="\t",row.names=FALSE,quote=FALSE)
```

# LD Analyses 

FDR 05 
```{r}

ld.dir <- "/Users/jtorres/FUSE/projects/wtsa/promoter_driven/" %&% "statistical_analysis/ld-proxies/"

build_ld_proxy_df <- function(){
  file.vec <- list.files(ld.dir)
  ld.files <- file.vec[grep("ld-proxies",file.vec)]
  out.df <- c()
  pb <- txtProgressBar(min=0,max=length(ld.files),style=3)
  for (i in 1:length(ld.files)){
    setTxtProgressBar(pb,i)
    f <- ld.files[i]
    df <- fread(ld.dir %&% f)
    out.df <- rbind(out.df,df)
  }
  return(out.df)
}

build_ld_proxy_gr <- function(ld.df){
  df1 <- select(ld.df,one_of("CHR1","POS1")); df1 <- df1[!duplicated(df1),]
  df2 <- select(ld.df,one_of("CHR2","POS2")); df2 <- df2[!duplicated(df2),]
  names(df1) <- c("CHR","POS");   names(df2) <- c("CHR","POS")
  df <- rbind(df1,df2); df <- df[!duplicated(df),]
  df$CHR <- "chr" %&% df$CHR
  gr <- GRanges(seqnames=df$CHR,IRanges(df$POS-1,df$POS))
  return(gr)
}

build_sensitivity_summary_ld <- function(my.fdr=0.05){
  eqtl.df <- as.data.frame(fread(ref.dir %&% "eqtl-index.bed"))
  eqtl.df <- eqtl.df[!duplicated(eqtl.df),]

  out.df <- c() 
  pb <- txtProgressBar(min=0,max=dim(eqtl.df)[1],style=3)
  for (i in 1:dim(eqtl.df)[1]){
    print(i)
    setTxtProgressBar(pb,i)
    sub.df <- eqtl.df[i,]
    eqtl.name <- sub.df$V4
    ld.df <- fread(ld.dir %&% eqtl.name %&% ".ld-proxies.txt")
    if (dim(ld.df)[1]>0){
      e.gr <- build_ld_proxy_gr(ld.df)
    } else{
      e.gr <- GRanges(seqnames=sub.df$V1,
                      IRanges(sub.df$V2,sub.df$V3),
                      mcols=sub.df$V4)      
    }
    gene <- strsplit(x=eqtl.name,split="_")[[1]][2]
    df1 <- get_subtract(gene,e.gr)
    df2 <- get_deseq(gene,e.gr)
    df2$deseq.sig <- (df2$deseq.padj > -log(0.05,base=exp(1)))
    df <- inner_join(df1,df2,by="id")

    df$peakCmod.win11.fdr05 <- sapply(sapply(1:dim(df)[1],
                                      function(i) {
                                        temp <- df$id[i]
                                        c <- strsplit(temp,":")[[1]][1]
                                        s <- as.integer(strsplit(strsplit(temp,":")[[1]][2],"-")[[1]][1])
                                        e <- as.integer(strsplit(strsplit(temp,":")[[1]][2],"-")[[1]][2])
                                        temp.gr <- GRanges(seqnames=c,IRanges(s,e))
                                        temp.e.gr <- e.gr[e.gr%over%temp.gr]
                                        get_peakC_mod(temp.e.gr,11,my.fdr)}),function(l){any(l)==TRUE})
    df$peakCmod.win60.fdr05 <- sapply(sapply(1:dim(df)[1],
                                      function(i) {
                                        temp <- df$id[i]
                                        c <- strsplit(temp,":")[[1]][1]
                                        s <- as.integer(strsplit(strsplit(temp,":")[[1]][2],"-")[[1]][1])
                                        e <- as.integer(strsplit(strsplit(temp,":")[[1]][2],"-")[[1]][2])
                                        temp.gr <- GRanges(seqnames=c,IRanges(s,e))
                                        temp.e.gr <- e.gr[e.gr%over%temp.gr]
                                        get_peakC_mod(temp.e.gr,60,my.fdr)}),function(l){any(l)==TRUE})
    df$peakCcomp.win11.cutoff140 <- sapply(sapply(1:dim(df)[1],
                                      function(i) {
                                        temp <- df$id[i]
                                        c <- strsplit(temp,":")[[1]][1]
                                        s <- as.integer(strsplit(strsplit(temp,":")[[1]][2],"-")[[1]][1])
                                        e <- as.integer(strsplit(strsplit(temp,":")[[1]][2],"-")[[1]][2])
                                        temp.gr <- GRanges(seqnames=c,IRanges(s,e))
                                        temp.e.gr <- e.gr[e.gr%over%temp.gr]
                                        get_peakC_comp(temp.e.gr,11,140)}),function(l){any(l)==TRUE})
    df$peakCcomp.win60.cutoff140 <- sapply(sapply(1:dim(df)[1],
                                      function(i) {
                                        temp <- df$id[i]
                                        c <- strsplit(temp,":")[[1]][1]
                                        s <- as.integer(strsplit(strsplit(temp,":")[[1]][2],"-")[[1]][1])
                                        e <- as.integer(strsplit(strsplit(temp,":")[[1]][2],"-")[[1]][2])
                                        temp.gr <- GRanges(seqnames=c,IRanges(s,e))
                                        temp.e.gr <- e.gr[e.gr%over%temp.gr]
                                        get_peakC_comp(temp.e.gr,60,140)}),function(l){any(l)==TRUE})
    df$peakCcomp.win11.cutoff10 <- sapply(sapply(1:dim(df)[1],
                                      function(i) {
                                        temp <- df$id[i]
                                        c <- strsplit(temp,":")[[1]][1]
                                        s <- as.integer(strsplit(strsplit(temp,":")[[1]][2],"-")[[1]][1])
                                        e <- as.integer(strsplit(strsplit(temp,":")[[1]][2],"-")[[1]][2])
                                        temp.gr <- GRanges(seqnames=c,IRanges(s,e))
                                        temp.e.gr <- e.gr[e.gr%over%temp.gr]
                                        get_peakC_comp(temp.e.gr,11,10)}),function(l){any(l)==TRUE})
    df$peakCcomp.win60.cutoff10 <- sapply(sapply(1:dim(df)[1],
                                      function(i) {
                                        temp <- df$id[i]
                                        c <- strsplit(temp,":")[[1]][1]
                                        s <- as.integer(strsplit(strsplit(temp,":")[[1]][2],"-")[[1]][1])
                                        e <- as.integer(strsplit(strsplit(temp,":")[[1]][2],"-")[[1]][2])
                                        temp.gr <- GRanges(seqnames=c,IRanges(s,e))
                                        temp.e.gr <- e.gr[e.gr%over%temp.gr]
                                        get_peakC_comp(temp.e.gr,60,10)}),function(l){any(l)==TRUE})[1:dim(df)[1]]
    sub.df <- data.frame(V1=rep(sub.df$V1,dim(df)[1]),
                         V2=rep(sub.df$V2,dim(df)[1]),
                         V3=rep(sub.df$V3,dim(df)[1]),
                         V4=rep(sub.df$V4,dim(df)[1]),
                         stringsAsFactors=FALSE)
    names(sub.df) <- c("chrom","pos0","pos","eqtl")
    df <- cbind(select(sub.df,one_of("chrom","pos","eqtl")),df)
    names(df)[grepl("id",names(df))] <- "frag.id"
    out.df <- rbind(out.df,df)
  }
  if (my.fdr==0.10){
    names(out.df)[grepl("fdr05",names(out.df))] <- c("peakCmod.win11.fdr10","peakCmod.win60.fdr10")
  } else if (my.fdr==0.20){
    names(out.df)[grepl("fdr05",names(out.df))] <- c("peakCmod.win11.fdr20","peakCmod.win60.fdr20")
  } else{
    print("FDR is at 0.05")
  }
  return(out.df)
}

get_sensitivity_significant_ld <- function(sens.ld.df,my.fdr=0.05){
  fdrname <- ifelse(my.fdr==0.05,"05",
                    ifelse(my.fdr==0.10,"10",
                           ifelse(my.fdr==0.20,"20",NA)))
  sig.vec <- c()
  for (i in 1:dim(sens.ld.df)[1]){
    print(i)
    #sig.vec <- append(sig.vec,any(na.omit(sens.ld.df[i,13:18])==TRUE))
    sig.vec <- append(sig.vec,any(na.omit(sens.ld.df[i,13:16])==TRUE))

  }
  sens.sig.df <- sens.ld.df[sig.vec,]
  sigsens <- length(unique(sens.sig.df$eqtl)) / length(unique(sens.ld.df$eqtl))
  print("Sensitivity: " %&% round(sigsens,3))
  sens.sig.df <- select(sens.sig.df,
                        one_of("chrom","pos","eqtl",
                                "frag.id","conD_mean","conT_mean",
                                "DIF_conD_minus_conT_mean",
                                "deseq.sig","peakCmod.win11.fdr"%&%fdrname,
                                "peakCmod.win60.fdr"%&%fdrname,
                                "peakCcomp.win11.cutoff140",
                                "peakCcomp.win60.cutoff140",
                               "peakCcomp.win11.cutoff10",
                                "peakCcomp.win60.cutoff10"))
  return(sens.sig.df)
}

sensLD05.df <- build_sensitivity_summary_ld(my.fdr=0.05)
sensLD05.sig.df <- get_sensitivity_significant_ld(sensLD05.df,my.fdr=0.05)
write.table(x=sensLD05.sig.df,file=stat.dir %&% "sensitivity-sig-table_LD_FDR05-proxies.txt",
            sep="\t",row.names=FALSE,quote=FALSE)

sensLD10.df <- build_sensitivity_summary_ld(my.fdr=0.10)
sensLD10.sig.df <- get_sensitivity_significant_ld(sensLD10.df,my.fdr=0.10)
write.table(x=sensLD10.sig.df,file=stat.dir %&% "sensitivity-sig-table_LD_FDR10-proxies.txt",
            sep="\t",row.names=FALSE,quote=FALSE)

sensLD20.df <- build_sensitivity_summary_ld(my.fdr=0.20)
sensLD20.sig.df <- get_sensitivity_significant_ld(sensLD20.df,my.fdr=0.20)
write.table(x=sensLD20.sig.df,file=stat.dir %&% "sensitivity-sig-table_LD_FDR20-proxies.txt",
            sep="\t",row.names=FALSE,quote=FALSE)

length(unique(sensLD05.sig.df$eqtl))/length(unique(sensLD05.df$eqtl)) # 0.50 sensitivity 
length(unique(sensLD10.sig.df$eqtl))/length(unique(sensLD10.df$eqtl)) # 0.50 sensitivity 
length(unique(sensLD20.sig.df$eqtl))/length(unique(sensLD20.df$eqtl)) # 0.50 sensitivity 



```


# Sensitivity range 


```{r}

w11.fdr05 <- sum(sens.df$peakCmod.win11.fdr05)/dim(sens.df)[1] # 0.09090909
w60.fdr05 <- sum(sens.df$peakCmod.win60.fdr05)/dim(sens.df)[1] # 0.1363636
w11.c140 <- sum(sens.df$peakCcomp.win11.cutoff140)/dim(sens.df)[1] # 0.2272727
w60.c140 <- sum(sens.df$peakCcomp.win60.cutoff140)/dim(sens.df)[1] # 0.2727273
w11.c10 <- sum(sens.df$peakCcomp.win11.cutoff10)/dim(sens.df)[1] # 0.3636364
w60.c10 <- sum(sens.df$peakCcomp.win60.cutoff10)/dim(sens.df)[1] # 0.4545455

# Combined without cutoff 10: 0.409
# Combined without cutoff 10: 0.545 

get_sens_ld <- function(name){ 
  neqtls <- length(unique(sensLD05.df$eqtl))
  log.vec <- c() 
  for (e in unique(sensLD05.df$eqtl)){
    #print(e)
    log <- any((filter(sensLD05.df,eqtl==e) %>% select(one_of(name)))==TRUE)
    log.vec <- append(log.vec,log)
  }
  sens <- sum(log.vec) / neqtls
  print("Sensitivity for " %&% name %&% " : " %&% sens)
}

get_sens_ld("peakCmod.win11.fdr05") # 0.136363636363636
get_sens_ld("peakCmod.win60.fdr05") # 0.181818181818182
get_sens_ld("peakCcomp.win11.cutoff140") # 0.272727272727273
get_sens_ld("peakCcomp.win60.cutoff140") # 0.272727272727273
get_sens_ld("peakCcomp.win11.cutoff10") # 0.636363636363636
get_sens_ld("peakCcomp.win60.cutoff10") # 0.727272727272727

# Combined without cutoff 10: 0.5
# Combined without cutoff 10: 0.818 

remain <- unique(sens.df$eqtl)[!(unique(sens.df$eqtl) %in% unique(sensLD05.sig.df$eqtl))]
#"rs10231021_DGKB"  "rs17168486_DGKB"  "rs7903146_TCF7L2" "rs8042275_LARP6" 

sub.vec <- c("rs4736819_NKX6-3","rs10231021_DGKB","rs17168486_DGKB","rs2165069_AP3S2",
             "rs2165069_C15orf38-AP3S2","rs2146105_WARS","rs11603334_STARD10",
             "rs11039165_MADD","rs11257658_CAMK1D","rs11708067_ADCY5")

sensLD05.sub.df <- sensLD05.df %>% filter(eqtl%in%sub.vec) 

get_sens_ld_sub <- function(name){ 
  neqtls <- length(unique(sensLD05.sub.df$eqtl))
  log.vec <- c() 
  for (e in unique(sensLD05.sub.df$eqtl)){
    #print(e)
    log <- any((filter(sensLD05.sub.df,eqtl==e) %>% select(one_of(name)))==TRUE)
    log.vec <- append(log.vec,log)
  }
  sens <- sum(log.vec) / neqtls
  print("Sensitivity for " %&% name %&% " : " %&% sens)
}

get_sens_ld_sub("peakCmod.win11.fdr05") # 0.1
get_sens_ld_sub("peakCmod.win60.fdr05") # 0.1
get_sens_ld_sub("peakCcomp.win11.cutoff140") # 0.2
get_sens_ld_sub("peakCcomp.win60.cutoff140") # 0.2
get_sens_ld_sub("peakCcomp.win11.cutoff10") # 0.55
get_sens_ld_sub("peakCcomp.win60.cutoff10") # 0.8

get_sens_ld_sub(c("peakCmod.win11.fdr05","peakCmod.win60.fdr05",
                  "peakCcomp.win11.cutoff140","peakCcomp.win60.cutoff140"))

```



# Gene Count analysis 

```{r}

ct.df <- fread(eda.dir %&% "cis-trans-summary.txt")

#Exclude cutoff 10 results 

eqtl.vec <- unique(sens.df$eqtl)
gene.vec <- sapply(i:length(eqtl.vec),function(i){
  eqtl <- eqtl.vec[i]
  s <- strsplit(eqtl,split="_")[[1]]
  s[length(s)]
}) 

#eqtl.sig.vec <-  c("rs11257658_CAMK1D","rs59153558_HMG20A", "rs3807042_KCNK17","rs113748381_POLR2A","rs2933243_RDH5",
#               "rs2146105_WARS","rs2910686_ERAP2","rs28642213_GPSM1","rs28679497_DNLZ",
 #              "rs61386106_CARD9","rs4736819_NKX6-3")

eqtl.sig.vec <- unique(filter(sensLD05.sig.df,peakCmod.win11.fdr05==TRUE | peakCmod.win60.fdr05==TRUE)$eqtl)
gene.sig.vec <- sapply(i:length(eqtl.sig.vec),function(i){
  eqtl <- eqtl.sig.vec[i]
  s <- strsplit(eqtl,split="_")[[1]]
  s[length(s)]
})

gene.nosig.vec <- gene.vec[!(gene.vec %in% gene.sig.vec)]

sub.df <- filter(ct.df,Gene %in% c(gene.sig.vec,gene.nosig.vec))

sub.df$Sig <- sub.df$Gene %in% gene.sig.vec

ggplot(data=filter(sub.df,Cell=="EndoB"),aes(x=Sig,y=Count)) + geom_boxplot() +
  theme_bw() + geom_jitter(shape=21,aes(fill=Sample),alpha=0.2) + 
    scale_y_continuous(name = "Number of unique reporter fragments",
                     breaks = seq(0,70000,5000),
                     limits=c(0,70000)) + xlab("eQTLs overlap significant interactions (Modeled)")

t.test(filter(sub.df,Cell=="EndoB",Sig==TRUE)$Count,filter(sub.df,Cell=="EndoB",Sig==FALSE)$Count) # p-value = 0.146

#	Welch Two Sample t-test

#data:  filter(sub.df, Cell == "EndoB", Sig == TRUE)$Count and filter(sub.df, Cell == "EndoB", Sig == FALSE)$Count
#t = 1.478, df = 47.378, p-value = 0.146
#alternative hypothesis: true difference in means is not equal to 0
#95 percent confidence interval:
# -1309.637  8567.683
#sample estimates:
#mean of x mean of y 
# 20165.37  16536.34 

```




#Additional analyses 

Genotype quick check 

```{r}

library("annotables")

geno.dir <- "/Users/jtorres/FUSE/projects/wtsa/promoter_driven/" %&% "statistical_analysis/genotypes/"
eqtl.df <- fread("/Users/jtorres/FUSE/projects/wtsa/promoter_driven/reference_files/eqtl-index.bed")

al.df <- fread(geno.dir%&%"eqtl-alleles.txt")
eref.df <- fread(geno.dir%&%"eqtl-summary.txt")
mpu.df <- fread(stat.dir%&%"bam_files/eqtls.mpileup")

append_symbols_to_eref <- function(){
  symbol <- c()
  for (i in 1:dim(eref.df)[1]){
    ensid <- strsplit(eref.df[i,]$Gene,split=".",fixed=TRUE)[[1]][1]
    sym <- unique(filter(grch37,ensgene==ensid)$symbol)
    if (length(sym)!=1){
      print(sym)
    }
    symbol <- append(symbol,sym)
  }
  out.df <- cbind(symbol,eref.df)
  out.df$chrom <- "chr" %&% out.df$CHROM
  return(out.df)
}

build_mpu_endo_df <- function(){
  mpu.endo.df <- select(mpu.df,one_of(c("V1","V2","V"%&%13:21)))
  mpu.endo.df$eqtl <- sapply(1:dim(mpu.endo.df)[1],
                             function(i){
                               sub.df <- mpu.endo.df[i,]
                               eqtl <- paste0(unique(filter(eqtl.df,V1==sub.df$V1,V3==sub.df$V2)$V4),
                                              collapse=":")})
  mpu.endo.df$ref <- sapply(1:dim(mpu.endo.df)[1],
                             function(i){
                               sub.df <- mpu.endo.df[i,]
                               ref <- paste0(unique(filter(al.df,CHROM==gsub("chr","",sub.df$V1),
                                                           POS==sub.df$V2)$REF),
                                              collapse=":")})
  mpu.endo.df$alt <- sapply(1:dim(mpu.endo.df)[1],
                             function(i){
                               sub.df <- mpu.endo.df[i,]
                               ref <- paste0(unique(filter(al.df,CHROM==gsub("chr","",sub.df$V1),
                                                           POS==sub.df$V2)$ALT),
                                              collapse=":")})  
  mpu.endo.df$gene <- sapply(1:dim(mpu.endo.df)[1],function(i){
    gene <- strsplit(mpu.endo.df$eqtl[i],split="_")[[1]][2]
    return(gene)
  })
  stack.df <- c()
  for (i in 1:dim(mpu.endo.df)[1]){
    temp <- mpu.endo.df[i,]
    if (grepl(":",temp$gene)){
      temp$gene <- strsplit(temp$gene,split=":")[[1]][1]
    }
    temp.df <- filter(eref.full.df,chrom==temp$V1,POS==temp$V2,symbol==temp$gene)
    sub.df <- select(temp.df,one_of("REF","ALT","P","Slope"))
    if (dim(sub.df)[1]==0){
      sub.df[1,] <- rep(NA,4)
    }
    print(dim(sub.df))
    stack.df <- rbind(stack.df,sub.df)
  }
  mpu.endo.df <- cbind(mpu.endo.df,stack.df)
  return(mpu.endo.df)
}


eref.full.df <- append_symbols_to_eref()
mpu.endo.df <- build_mpu_endo_df()



s <- mpu.endo.df[1,]
ss <- filter(eref.df,CHROM==gsub("chr","",s$V1),POS==s$V2)

```




