---
title: "Untitled"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Setup 

```{r}
"%&%" <- function(a,b) paste0(a,b)
library("data.table");library("tidyverse")
library("GenomicRanges")
got2d.dir <- "/home/jason/science/servers/FUSE/"
serv.dir <- "/home/jason/science/servers/FUSE5/"
work.dir <- serv.dir %&% "projects/wtsa/enhancer_driven/first_round/fgwas_probe_design/first_round/output_files/"

```


```{r}

allsnp.df <- fread(work.dir %&% "All_SNPs_of_interest.csv")#newOligofile_shortNames_modified.txt
unique(allsnp.df$ID) %>% length(.)
unique(allsnp.df$SEGNUMBER) %>% length(.)
unique(allsnp.df$SEGNUMBER) %>% map(.,function(id){grepl("QTL",id)}) %>% as.logical(.) %>% sum(.)
unique(allsnp.df$SEGNUMBER) %>% map(.,function(id){grepl("TG",id)}) %>% as.logical(.) %>% sum(.)

allsnp.df$snp <- map(allsnp.df$ID,function(id){
  strsplit(x=id,split=",")[[1]][4]
}) %>% as.character(.)

allsnp.df$snp %>% unique(.) %>% length(.) # 283 unique SNPs 
sub.df <- dplyr::select(allsnp.df,one_of("Capture.Status","Fragment"))
sub.df <- sub.df[!duplicated(sub.df),]
table(sub.df$Capture.Status)



```


# Get SNP counts 


```{r}

outfile.dir <- got2d.dir %&% "projects/wtsa_archive/wtsa/enhancer_driven/" %&%
  "fgwas_probe_design/first_round/"

df <- readRDS(outfile.dir%&%"RDS/first.round.full.df.RDS")
sdf <- filter(df,SNPID%in%allsnp.df$snp)
sdf <- filter(sdf, !grepl("offset",SNPID))
filter(allsnp.df,!(snp %in% sdf$SNPID))$ID
dup.vec <- sdf$SNP %>% duplicated(.)
sdf[dup.vec,] # DUPLICATED SNPS 
cv.df <- filter(sdf,grepl("DF",SEGNUMBER) | SEGNUMBER %in%
                  c("9QTL")) # "1QTL","4QTL","6TG", "12QTL", ,"5TG" is redundant 
cv.df$SNPID %>% unique(.) %>% length(.) # 248 unique SNPs
cv.df$SEGNUMBER %>% unique(.) %>% length(.) # 26 loci 
filter(cv.df,Type=="upstream.control")$SEGNUMBER %>% unique(.) %>% length(.) # 24 loci 
filter(cv.df,Type=="downstream.control")$SEGNUMBER %>% unique(.) %>% length(.) # 25 loci 


compdf <- readRDS(outfile.dir%&%"RDS/round1.loci.df.RDS")



```



```{r}

pruned.df <- fread(work.dir %&% "Pruned_oligo_probeset.csv")
sub.df <- dplyr::select(pruned.df,one_of("type","quality","fragment"))
sub.df <- sub.df[!duplicated(sub.df),]
filter(sub.df,type=="paired")$quality %>% table(.)
filter(sub.df,type=="single")$quality %>% table(.)

```



# Output from bioinformatic capture anlaysis (CCAnalyser CM5)

```{r}

cap.df <- fread(work.dir %&% "newOligofile_shortNames_modified.txt")
fail.df <-  fread(work.dir %&% "failed_captures_list.txt", header=F)
fail.df$id <- map(fail.df$V1,function(s){
  r <- regexpr(pattern="chr\\d+_",s)
  i <- attr(r,"match.length")
  substr(s,start=i+1,stop=nchar(s))  
}) %>% as.character(.)

fail.df$id %in% cap.df$V1 %>% sum(.)

good.df <- filter(cap.df,!(V1 %in% fail.df$id))
good.df$V2 <- "chr" %&% good.df$V2

good.gr <- GRanges(seqnames = good.df$V2, IRanges(start=good.df$V3,end=good.df$V4))

allsnp.df$success <- map(1:dim(allsnp.df)[1],function(i){
  row.df <- allsnp.df[i,]
  snp.gr <- GRanges(seqnames = row.df$Chrom, IRanges(start=row.df$Start,end=row.df$End))
  snp.gr %over% good.gr
}) %>% as.logical(.)

goodsnp.df <- filter(allsnp.df,success==TRUE)
goodsnp.df$snp %>% unique(.) %>% length(.)

```

I found out that only 191 of the 238 fragments were actually captured in the experiment whereas JT mistakenly included all 238 fragments in the oligofile for CCanalyser 


```{r}
redund.vec <- c("1QTL","3QTL","4QTL","6TG","12QTL","5TG")
success.df <- dplyr::select(goodsnp.df,one_of("snp","SEGNUMBER")) %>% 
  filter(.,!(SEGNUMBER %in% redund.vec))
success.df <- success.df[!duplicated(success.df),] # 257 snps 
success.df$SEGNUMBER %>% unique(.) %>% length(.)                          

# "9QTL" ,"1QTL","4QTL","6TG", "12QTL", ,"5TG" is redundant 


```

