---
title: "Untitled"
author: "Jason Torres"
date: "February 15, 2017"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r}
"%&%" <- function(a,b) paste0(a,b) 
library("data.table")
library("dplyr")
library("GenomicRanges")

serv.dir <- "/Users/jtorres/FUSE/"
rds.dir <- serv.dir %&% "projects/wtsa/enhancer_driven/probe_design/rds/"

```

Read RDS files 

```{r}

credt2d.df <- readRDS(rds.dir%&%"credt2d.df.RDS")
eqtl.df <- readRDS(rds.dir%&%"eqtl.df.RDS")
atac.df <- readRDS(rds.dir%&%"atac.df.RDS")

```

Islet eQTLs in T2D credible regions 

```{r}
 
eqt2d.dir <- serv.dir %&% "reference/islet/eqtls/from_anthony/t2d_credible_regions/"
eqt2d.df <- fread("cat " %&% eqt2d.dir %&% 
                  "cis_eQTL_genelevel-t2d-credible.txt.gz" %&% " | zmore")

```



Read in DpnII file 
Read 0-based "bed" file of DpnII (hg19) cut sites (from Jelena Telenius )

```{r}
dpn.file <- serv.dir %&% "reference/DpnII/hg19_DpnII-sites.bed.gz" #  0-based bed file 
dpn.df <- fread("cat " %&% dpn.file %&% " | zmore")
# adjust to 1-based scheme 
dpn.df$V2 <- dpn.df$V2 + 1 
dpn.df$V3 <- dpn.df$V3 + 1 
```


Passed probes from Martijn 

```{r}

prob.dir <- serv.dir %&% "projects/wtsa/enhancer_driven/" %&% "probe_design/from_Martijn/DpnII_120bp/"
probe.df <- fread(prob.dir %&% "lessThan100_all_success.txt")

```



Define functions for reading in input (bed) files into grange objects and checking for overlaps 

```{r}

gread <- function(bed.df){
  # names must be c(chr, start, end, id, score, strand), where available 
   if(length(bed.df)==3){
      gr <- with(bed.df, GRanges(chr, IRanges(start, end)))
   } else if (length(bed.df)==4){
      gr <- with(bed.df, GRanges(chr, IRanges(start, end), id=id))
   } else if (length(bed.df)==5){
      gr <- with(bed.df, GRanges(chr, IRanges(start, end), id=id, score=score))
   } else if (length(bed.df)>=6){
      bed.df$strand <- gsub(pattern="[^+-]+",replacement='*', bed.df$strand)
      gr <- with(bed.df, GRanges(chr, IRanges(start, end), id=id, score=score, strand=strand))
   }
   return(gr)
}

get_overlaps <- function(a,b){
  # a and b are genomic ranges
  vec <- a %over% b
  ##print(sum(vec) %&% " / " %&% length(vec))
  return(vec)
}


```


# Functions for creating ggplot objects 


**EndoC-BH1** 

```{r}

atac_plot <- function(segnum,fac=0.10){
  df <- filter(credt2d.df,SEGNUMBER==segnum)
  mx <- max(df$POS); mn <- min(df$POS); span <- mx - mn
  chrom <- df$CHR[1];mymin <- (mn-fac*span);mymax <- (mx+fac*span)
  #loc <- df$LOCUS[1]
  a1sub <- filter(atac.df,chr==chrom,start>=mymin,end<=mymax,id=="Oxford") %>% 
    dplyr::select(one_of("chr","start","end"))
  a1.g <- gread(a1sub[,1:3])
  a2sub <- filter(atac.df,chr==chrom,start>=mymin,end<=mymax,id=="Parker") %>% 
    dplyr::select(one_of("chr","start","end"))
  a2.g <- gread(a2sub[,1:3])
  if (length(a1.g)==0){
    a1.g <- GRanges(seqnames=chrom,IRanges(start=0,end=0),id="null")
  }
  if (length(a2.g)==0){
    a2.g <- GRanges(seqnames=chrom,IRanges(start=0,end=0),id="null")
  }
  plt <- ggplot(a2.g) + geom_segment(color="firebrick3",size=30,alpha=0.9) +
    geom_segment(a1.g,color="yellow2",size=25,alpha=0.9) + ylab("") +
    theme_bw() + theme(panel.grid = element_blank())
  return(plt)
}

```

**DpnII**

```{r}

dpnII_plot <- function(segnum,fac=0.10){
  df <- filter(credt2d.df,SEGNUMBER==segnum)
  mx <- max(df$POS); mn <- min(df$POS); span <- mx - mn
  chrom <- df$CHR[1];mymin <- (mn-fac*span);mymax <- (mx+fac*span)
  dpnsub <- filter(dpn.df,V1==chrom,V2>mymin,V3<mymax)
  if (dim(dpnsub)[1]==0){
    dpnsub <- data.frame(V1="chr0",V2=NA,V3=NA)
  }
  plt <-  ggplot(data=dpnsub) +
    geom_vline(aes(xintercept=V2), size=0.1) + 
    geom_vline(aes(xintercept=V3),size=0.1) + ylim(c(0,1)) + 
    theme_bw() + theme(axis.text.y=element_blank(),
                       axis.ticks = element_blank(),
                       panel.grid = element_blank())  
  return(plt)
}

```


**Credible Sets: LD**

```{r}

credld_plot <- function(segnum,fac=10){
  df <- filter(credt2d.df,SEGNUMBER==segnum)
  mx <- max(df$POS); mn <- min(df$POS); span <- mx - mn
  chrom <- df$CHR[1];mymin <- (mn-fac*span);mymax <- (mx+fac*span)  
}


```


**Credible Sets: fGWAS** 

```{r}

fgwas_plot <- function(segnum,fac=0.10){
  df <- filter(credt2d.df,SEGNUMBER==segnum)
  mx <- max(df$POS); mn <- min(df$POS); span <- mx - mn
  chrom <- df$CHR[1];mymin <- (mn-fac*span);mymax <- (mx+fac*span)
  plt <- ggplot(df,aes(x=POS)) + 
                geom_point(aes(y=PPA),shape=21,
                           color="gray",fill="gray") + 
                geom_point(aes(y=PPA.fgwas,fill=(change>0)),
                           shape=23,color="black") +   
                scale_fill_manual(values=c("dodgerblue2","firebrick2")) + 
    xlab("Position on Chromosome " %&% chrom) + 
    theme_bw() +  
    theme(legend.position="none",
          panel.grid.minor=element_blank(),
                axis.title.y=element_text(size=5),
                axis.text.y=element_text(size=4)) +
    scale_y_continuous(breaks=seq(0,1,0.1),limits=c(0,1)) + 
    coord_cartesian(ylim=c(0,1),expand=TRUE)
  return(plt)
}

```


**Gene Plot**

```{r}

library(ggbio)
library(Homo.sapiens)
gene_plot <- function(segnum,fac=0.10){
  df <- filter(credt2d.df,SEGNUMBER==segnum)
  mx <- max(df$POS); mn <- min(df$POS)
  chrom <- df$CHR[1]; span <- mx - mn
  gr <- GRanges(chrom,IRanges(start=mn-1e6,end=mx+1e6))
  aplt <- autoplot(Homo.sapiens, which = gr, label.color = "black",#
                 color = "brown",
                 fill = "brown") + #,stat="reduce"
        xlim(c(mn,mx)) + 
        theme_alignment(grid = FALSE,border = FALSE) 
  return(aplt)
}

```


**Islet eQTL** 

```{r}

islet_eqtl_plot <- function(segnum,fac=0.10){
  df <- filter(credt2d.df,SEGNUMBER==segnum)
  mx <- max(df$POS); mn <- min(df$POS)
  chrom <- df$CHR[1]; loc <- df$LOCUS[1]; span <- mx - mn
  mymin <- (mn-fac*span); mymax <- (mx+fac*span)
  sub.df <- filter(eqt2d.df,chr==chrom,pos>=mymin) %>% filter(pos<=mymax)
  e.sub <- filter(eqtl.df, CHR==chrom, POS>=mymin) %>% filter(POS<=mymax)
  if (dim(sub.df)[1]==0){
    sub.df <- data.frame(snp=NA,chr=NA,pos=0,gene=NA,beta=NA,tstat=NA,
                         pval=1,qval=NA)
  }
  if (dim(e.sub)[1]==0){
    e.sub <- data.frame(RSID=NA,ENSID=NA,Beta=NA,`T`=NA,P=1,Q=NA,CHR=NA,
                        POS=0,GENE=NA,SNPID=NA)
  }
  plt <- ggplot(data=e.sub, aes(x=POS,y=-log(P,base=10))) +
                geom_point(data=sub.df,aes(x=pos,y=-log(pval,base=10)),
                           shape=21,color="grey", size=2,
                           fill="grey",alpha=0.80) + 
                geom_point(color="black",shape=21,stroke=1.1,
                           fill="green1",size=2) + 
          ylab(expression(paste("-log"[10],"(p-value)"))) + theme_bw() + 
          theme(panel.grid = element_blank(),
                axis.title.y=element_text(size=5),
                axis.text.y=element_text(size=4)) +  xlim(c(mn,mx))
  return(plt)
}
islet_eqtl_plot(1)

```


## Tracks Function 

```{r}

track_plot <- function(segnum,fac=0.10){
  df <- filter(credt2d.df,SEGNUMBER==segnum)
  mx <- max(df$POS); mn <- min(df$POS); span <- mx - mn
  
  atac_plt <- atac_plot(segnum,fac)
  dpn_plt <- dpnII_plot(segnum,fac)
  fgwas_plt <- fgwas_plot(segnum,fac)
  eqtl_plt <- islet_eqtl_plot(segnum,fac)
  gene_plt <- gene_plot(segnum,fac)
  
  tracks(`ATAC \nEndoC`=atac_plt,
          `Islet eQTL`=eqtl_plt,
          `DpnII`=dpn_plt,
          `T2D Credible Set \n(Islet fGWAS)`=fgwas_plt,
          `Genes`=gene_plt,
          heights=c(1,1,1,3,1),title=loc %&% "\nInterval: " %&% 
           round(span/1000,digits=2) %&%" kb",
         label.text.cex=0.7,main.height=2) +
    scale_x_sequnit("Mb") + 
    theme(axis.text.x=element_text(size=5))
}

```



```{r}



#viz_loc(20)

```

