---
title: "Evaluate MIG Exported Oligos file for filtering "
author: "Jason Torres"
date: "June 8, 2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# Setup 
```{r}

"%&%" <- function(a,b) paste0(a,b) 
library("data.table")
library("dplyr")

serv.dir <- "/Users/jtorres/FUSE/"
work.dir <- serv.dir %&% "projects/wtsa/enhancer_driven/fgwas_probe_design/first_round/"

#rds.dir <- work.dir %&% "RDS/"
txt.dir <- work.dir %&% "txt/"
cap.dir <- work.dir %&% "capsequm_output/"

roi.df <- fread(txt.dir %&% "first-round-variants_error-adjust.bed",header=F,sep="\t")
mig.df <- fread(cap.dir %&% "MIG_exported_data.txt") 


```



# Correct FragIDs 

```{r}

fix_mig_ids <- function(){
  out.df <- mig.df
  FragID <- c()
  nums <- c(1,2,3,4)
  for (i in 1:dim(mig.df)[1]){
    vec <- strsplit(split=", ",mig.df$FragID[i])[[1]]
    snpid <- vec[grepl(pattern=":",vec) & !(grepl(pattern="_",vec))]
    segnum <- vec[grepl("DF",vec)|grepl("TG",vec)|grepl("QTL",vec)]
    segnum <- segnum[!grepl(":",x=segnum)]
    segnum <- segnum[!grepl("eQTL",x=segnum)]
    type <- vec[grepl("DIAMANTE.Morris",vec)|grepl(".",vec,fixed=TRUE)|grepl("eQTL",vec)]
    type <- type[!grepl(":",x=type)]
    index <- nums[!(nums %in% c(match(snpid,vec),match(segnum,vec),match(type,vec)))]
    name <- strsplit(split="_",x=vec[index])[[1]]
    name <- paste(name[2:length(name)],sep="_",collapse="_")
    fragid <- paste(name,segnum,type,snpid,sep=",")
    print(c(i,fragid))
    FragID <- c(FragID,fragid)
  }
  out.df$FragID <- FragID
  return(out.df)
}

mig.df <- fix_mig_ids()

```


Manually Remove ACSL1 gene variants (33 is too many)

```{r}

roi.df <- filter(roi.df,!grepl("ACSL1",roi.df$V4))
mig.df <- filter(mig.df,!grepl("ACSL1",mig.df$FragID))

saveRDS(mig.df,cap.dir%&%"mig.df.RDS")



```



# Filter oligos 

```{r}

get_filter_df <- function(){
  thresh <- 70 * 0.25
  filter.df <- filter(mig.df,Density<=30)
  out.df <- c()
  for (i in 1:dim(filter.df)[1]){
    row <- filter.df[i,]
    if (filter.df$SRepeatLength <= thresh | filter.df$Percent_CG >= 0.50 | filter.df$Percent_GC >= 0.50){
      out.df <- rbind(out.df,row)
    } else{
      print(row)
    }
  }
  return(out.df)
}

get_paired_df <- function(){
  filter.df <- get_filter_df()
  filter.df$chrom <- as.integer(gsub("chr","",filter.df$chr))
  oligo.df <- arrange(filter.df,chrom,`bp start`,FragID)
  out.df <- c()
  for (frag in unique(oligo.df$FragID)){
    stack.df <- filter(oligo.df,FragID==frag)
    if (dim(stack.df)[1]==2){
      out.df <- rbind(out.df,stack.df)
    }
  }
  return(out.df)   
}

paired.df <- get_paired_df()
saveRDS(paired.df,cap.dir%&%"high.quality.paired.probes.df.RDS")
write.table(x=paired.df,file=cap.dir%&%"high-quality-oligos_first-round_enhancer-drivenCapC.txt",
            sep="\t",quote=FALSE,row.names=FALSE)


get_frag_df <- function(paired.df){
  frag <- c()
  id <- c() 
  chrom <- c()
  start <- c()
  end <- c() 
  for (i in 1:length(unique(paired.df$FragID))){
    frag <- c(frag,"frag"%&%i)
    f<- unique(paired.df$FragID)[i]
    id <- c(id,f)
    stack.df <- filter(paired.df,FragID==f) %>% arrange(`bp start`)
    ch <- stack.df$chr[1]
    chrom <- c(chrom,ch)
    start <- c(start,stack.df$`bp start`[1])
    end <- c(end,stack.df$`bp end`[2])
  }
  out.df <- data.frame(frag,id,chrom,start,end,stringsAsFactors = FALSE)
  return(out.df) 
}
frag.df <- get_frag_df(paired.df)
saveRDS(frag.df,cap.dir%&%"high.quality.frags.df.RDS")
write.table(x=paired.df,file=cap.dir%&%"high-quality-fragments_first-round_enhancer-drivenCapC.txt",
            sep="\t",quote=FALSE,row.names=FALSE)

```


# Medium-high quality (Density <= 40, single or paired)

```{r}

remain.df <- filter(mig.df,!(FragID %in% paired.df$FragID)) %>% arrange(as.integer(gsub("chr","",chr)),FragID,`bp start`)
remain.tg.df <- filter(remain.df,grepl("DIAMANTE.Morris",remain.df$FragID))
remain.qtl.df <- filter(remain.df,grepl("rs",remain.df$FragID))

medhigh.df <- filter(remain.df, Density<=40) # changed to 40 
get_mh_df <- function(medhigh.df){
  filter.df <- medhigh.df
  filter.df$chrom <- as.integer(gsub("chr","",filter.df$chr))
  oligo.df <- arrange(filter.df,chrom,`bp start`,FragID)
  out.df <- c()
  for (frag in unique(oligo.df$FragID)){
    stack.df <- filter(oligo.df,FragID==frag)
    #if (dim(stack.df)[1]==2){ # no longer require paired probes 
    #  out.df <- rbind(out.df,stack.df)
    #}
    out.df <- rbind(out.df,stack.df)
  }
  return(out.df)   
}

medhigh.df <- get_mh_df(medhigh.df)
saveRDS(medhigh.df,cap.dir%&%"medium.quality.probes.df.RDS")
write.table(x=medhigh.df,file=cap.dir%&%"medium-quality-oligos_first-round_enhancer-drivenCapC.txt",
            sep="\t",quote=FALSE,row.names=FALSE)

#med.frag.df <- get_frag_df(medhigh.paired.df)
#saveRDS(med.frag.df,cap.dir%&%"medium.quality.frags.df.RDS")
#write.table(x=med.frag.df,file=cap.dir%&%"medium-quality-fragments_first-round_enhancer-drivenCapC.txt",
#            sep="\t",quote=FALSE,row.names=FALSE)

# Get single high quality probes 
#rem.df <-  filter(remain.df,!(FragID %in% medhigh.paired.df$FragID)) %>% arrange(as.integer(gsub("chr","",chr)),FragID,`bp start`)
#single.high.df <- filter(rem.df,Density<=30)
#saveRDS(single.high.df,cap.dir%&%"high.quality.single.probes.df.RDS")
#write.table(x=single.high.df,file=cap.dir%&%"high-quality-SINGLE-oligos_first-round_enhancer-drivenCapC.txt",
#            sep="\t",quote=FALSE,row.names=FALSE)
```


# Filter out probes that capture redundant fragments 

Read in DpnII file 
Read 0-based "bed" file of DpnII (hg19) cut sites (from Jelena Telenius )

```{r}
dpn.file <- serv.dir %&% "reference/DpnII/hg19_DpnII-sites.bed.gz" #  0-based bed file 
dpn.df <- fread("cat " %&% dpn.file %&% " | zmore")
# adjust to 1-based scheme 
dpn.df$V2 <- dpn.df$V2 + 1 
dpn.df$V3 <- dpn.df$V3 + 1 

dpn.df$fragid <- dpn.df$V1%&%":"%&%dpn.df$V2%&%"-"%&%dpn.df$V3
```



```{r}

medhigh.df$chrom <- as.integer(gsub("chr","",medhigh.df$chr))
full.df <- rbind(paired.df,medhigh.df)

# Add manual entries 
sub.df <- as.data.frame(matrix(rep(NA,19),nrow=1,ncol=19))
names(sub.df) <- names(full.df)
row1 <- c("chr3",187741980,187742050,TRUE,34.61,1,FALSE, "LOC105374266,29DF,credible.variant,chr3:187741842",70,0,30,NA,NA,NA,	"GTTTTAAGTGGAAATTTTAAAATATGTGCATACATGTGAGACTTCTGAGTCCTATCATTCATATTTGATC",FALSE,0,NA,3)
sub.df[1,] <- row1

row2 <- c("chr10",	12308720,	12308790,	TRUE,	21.27,	1,	FALSE, "CDC123,5TG,DIAMANTE.Morris,chr10:12307894.offset1000", 70	,0	,22.86,	NA,	NA,	NA,	"GATCAAAGCTTAATATAGTTAACATACAAATAAAAGGACTTCAAAATTAAACAAAATAAGATTGGGTTAA",	FALSE,	0, NA,10)
sub.df[2,] <- row2

row3 <- c("chr6",20689124,20689194,TRUE,14.26,4,TRUE,"CDKAL1,50DF,credible.variant,chr6:20688121.offset50",70,0,47.14,NA,NA,NA,"GGGAGAGGGAGTGTTCTGTAATCCACTGATTAGGTCTCAGTCCTTTAGTAAGCCTGTGCTTACTGTGATC",	FALSE,0,NA,6)
sub.df[3,] <- row3

full.df <- rbind(full.df,sub.df)
full.df$`bp start` <- as.integer(full.df$`bp start`)
full.df$`bp end` <- as.integer(full.df$`bp end`)
full.df$chrom <- as.integer(full.df$chrom)

# Filter outprobes that capture redundant fragments  

full.df <- arrange(full.df,chrom,`bp start`)

filter_redun <- function(){
  track.vec <- c()
  type <- c() 
  quality <- c()
  fragment <- c() 
  out.df <- c()
  for (id in unique(full.df$FragID)){
    temp.df <- filter(full.df,FragID==id)
    temp.df$Density <- as.integer(temp.df$Density)
    temp.df <- arrange(temp.df,`bp start`)
    c <- unique(temp.df$chr)
    if (dim(temp.df)[1]==2){
      pos1 <- temp.df$`bp start`[1]
      pos2 <- temp.df$`bp end`[2]
      f <- filter(dpn.df,V1==c,V2>=pos1,V3<=pos2)
    } else if(regexpr(pattern="GATC",text = temp.df$Sequence) == 1){
      pos <- temp.df$`bp start`
      f <- filter(dpn.df,V1==c,V2==(pos+2))
    } else if(regexpr(pattern="GATC",text = temp.df$Sequence) > 1){
      pos <- temp.df$`bp end`
      f <- filter(dpn.df,V1==c,V3==(pos-2))
    }
    #; pos <- temp.df$`bp start`[1]
    #f <- filter(dpn.df,V1==c,V2<=pos,V3>=pos)
    fragname <- f$V1 %&% ":" %&% f$V2 %&% "-" %&% f$V3
    if (!(fragname %in% track.vec)){
      out.df <- rbind(out.df,temp.df)
      track.vec <- c(track.vec,fragname)
      if (dim(temp.df)[1] ==2){
        fragment <- c(fragment,rep(fragname,2))
        type <- c(type,rep("paired",2))
        if (all(temp.df$Density <=30)){
          quality <- c(quality,rep("HighQuality",2))
        } else{
          quality <- c(quality,rep("MediumQuality",2))
        }         
      } else if(dim(temp.df)[1]==1){
        fragment <- c(fragment,fragname)
        type <- c(type,"single")
        if (all(temp.df$Density <=30)){
          quality <- c(quality,"HighQuality")
        } else{
          quality <- c(quality,"MediumQuality")
        }  
      } else {
        print("Please inspect")
      }
    }
  }
  out.df <- cbind(out.df,type,quality,fragment)
  out.df$type <- as.character(out.df$type)
  out.df$quality <- as.character(out.df$quality)
  out.df$fragment <- as.character(out.df$fragment)
  return(out.df)
}

filtered.df <- filter_redun()
write.table(x=filtered.df,file=cap.dir%&%"NonRedundant-Oligo-Probeset.txt",
            sep="\t",quote=FALSE,row.names=FALSE)
```


# Prune adjacent fragments from being captured 

```{r}

prune <- function(filtered.df){
  out.df <- c() 
  inspect.vec <- filtered.df$fragment
  while(length(inspect.vec)>0){
    print(length(inspect.vec))
    f <- inspect.vec[1]
    c <- strsplit(x=f,split=":")[[1]][1]
    pos <- strsplit(x=f,split=":")[[1]][2]
    start <- as.integer(strsplit(x=pos,split="-")[[1]][1])
    end <- as.integer(strsplit(x=pos,split="-")[[1]][2])
    index <- match(f,dpn.df$fragid)
    f.adj.up <- dpn.df[(index-1),]$fragid
    f.adj.dn <- dpn.df[(index+1),]$fragid
    stack.df <- filter(filtered.df,fragment==f)
    out.df <- rbind(out.df,stack.df)
    remove.vec <- c(f.adj.up,f,f.adj.dn)
    inspect.vec <- inspect.vec[!(inspect.vec %in% remove.vec)]
  }
  return(out.df)
}


pruned.df <- prune(filtered.df)
write.table(x=pruned.df,file=cap.dir%&%"Pruned-Oligo-Probeset.txt",
            sep="\t",quote=FALSE,row.names=FALSE)

```


# Determine if variants will be covered 


```{r}

annot_roi <- function(){
  capturable <- c()
  segid <- c() 
  fragment <- c()
  pb <- txtProgressBar(min=0,max=dim(roi.df)[1],initial=0,style = 3)
  for (i in 1:dim(roi.df)[1]){
    setTxtProgressBar(pb,i)
    ch <- roi.df$V1[i]
    pos <- roi.df$V3[i]
    varid <- roi.df$V4[i]
    seg <- strsplit(roi.df$V4[i],split=",")[[1]][2]
    segid <- c(segid,seg)
    frag <- filter(dpn.df,V1==ch,V2<=pos,V3>=pos)$fragid
    fragment <- c(fragment,frag)
    if (frag %in% filtered.df$fragment){
      temp.df <- filter(filtered.df,fragment==frag)
      temp.df$capt <- temp.df$quality %&% "." %&% temp.df$type
      capturable <- c(capturable,unique(temp.df$capt))
    } else{
      capturable <- c(capturable,"PoorQuality")
    }
    #h.frag.sub <- filter(frag.df,chrom==ch,start<=pos & end >=pos)
    #m.frag.sub <- filter(med.frag.df,chrom==ch,start<=pos & end >=pos)
    
    #if (dim(h.frag.sub)[1]>0){
    #  capturable <- c(capturable,"highQuality.PairedProbes")
    #} else if(dim(m.frag.sub)[1]>0){
    #  capturable <- c(capturable,"mediumQuality.PairedProbes")
    #} else if (varid %in% single.high.df$FragID){
    #  capturable <- c(capturable,"highQuality.SingleProbe")
    #} else{
    #  capturable <- c(capturable,"PoorQuality")
    #}
  }
  
  out.df <- cbind(roi.df,capturable,segid,fragment) # 289 variants 
  names(out.df) <- c("Chrom","Start","End","ID","Capture.Status","SEGNUMBER","Fragment")
  out.df$Capture.Status <- as.character(out.df$Capture.Status)
  out.df$SEGNUMBER <- as.character(out.df$SEGNUMBER)
  out.df$Fragment <- as.character(out.df$Fragment)
  return(out.df)
  
}

roi.cov.df <- annot_roi()

saveRDS(roi.cov.df,cap.dir%&%"roi.df.RDS")
write.table(x=roi.cov.df,file=cap.dir%&%"SNPs-of-interest_enhancer-drivenCapC.txt",
            sep="\t",quote=FALSE,row.names=FALSE)

table(roi.cov.df$Capture.Status)

capture.df <- filter(roi.cov.df,Capture.Status!="PoorQuality") # 253 
poor.df <- filter(roi.cov.df,Capture.Status=="PoorQuality") # 36

write.table(x=poor.df,file=cap.dir%&%"PoorQuality_SNPs-of-interest_enhancer-drivenCapC.txt",
            sep="\t",quote=FALSE,row.names=FALSE)

```



Profile fragment redundancies 

```{r}

gread <- function(bed.df){
  # names must be c(chr, start, end, id, score, strand), where available 
   if(length(bed.df)==3){
      gr <- with(bed.df, GRanges(chr, IRanges(start, end)))
   } else if (length(bed.df)==4){
      gr <- with(bed.df, GRanges(chr, IRanges(start, end), id=id))
   } else if (length(bed.df)==5){
      gr <- with(bed.df, GRanges(chr, IRanges(start, end), id=id, score=score))
   } else if (length(bed.df)>=6){
      bed.df$strand <- gsub(pattern="[^+-]+",replacement='*', bed.df$strand)
      gr <- with(bed.df, GRanges(chr, IRanges(start, end), id=id, score=score, strand=strand))
   }
   return(gr)
}

get_overlaps <- function(a,b){
  # a and b are genomic ranges
  vec <- a %over% b
  ##print(sum(vec) %&% " / " %&% length(vec))
  return(vec)
}


```


